src/NHibernate/Linq/Visitors/ParameterTypeLocator.cs

```diff
@@ -282,41 +281,53 @@ protected override Expression VisitConstant(ConstantExpression node)
 			}
 
 			protected override Expression VisitSubQuery(SubQueryExpression node)
+			{
+				if (!TryLinkContainsMethod(node.QueryModel))
+				{
+					node.QueryModel.TransformExpressions(Visit);
+				}
+
+				return node;
+			}
+
+			private bool TryLinkContainsMethod(QueryModel queryModel)
 			{
 				// ReLinq wraps all ResultOperatorExpressionNodeBase into a SubQueryExpression. In case of
 				// ContainsResultOperator where the constant expression is dislocated from the related expression,
 				// we have to manually link the related expressions.
-				if (node.QueryModel.ResultOperators.Count == 1 &&
-				    node.QueryModel.ResultOperators[0] is ContainsResultOperator containsOperator &&
-				    node.QueryModel.SelectClause.Selector is QuerySourceReferenceExpression querySourceReference &&
-				    querySourceReference.ReferencedQuerySource is MainFromClause mainFromClause &&
-				    mainFromClause.FromExpression is ConstantExpression constantExpression)
+				if (queryModel.ResultOperators.Count != 1 ||
+					!(queryModel.ResultOperators[0] is ContainsResultOperator containsOperator) ||
+					!(queryModel.SelectClause.Selector is QuerySourceReferenceExpression querySourceReference) ||
+					!(querySourceReference.ReferencedQuerySource is MainFromClause mainFromClause))
 				{
-					VisitConstant(constantExpression);
-					AddRelatedExpression(constantExpression, UnwrapUnary(Visit(containsOperator.Item)));
-					// Copy all found MemberExpressions to the constant expression
-					// (e.g. values.Contains(o.Name != o.Name2 ? o.Enum1 : o.Enum2) -> copy o.Enum1 and o.Enum2)
-					if (RelatedExpressions.TryGetValue(containsOperator.Item, out var set))
-					{
-						foreach (var nestedMemberExpression in set)
-						{
-							AddRelatedExpression(constantExpression, nestedMemberExpression);
-						}
-					}
+					return false;
+				}
+
+				var left = UnwrapUnary(mainFromClause.FromExpression);
+				var right = UnwrapUnary(containsOperator.Item);
+				if (left.NodeType == ExpressionType.Constant)
```

Comment:
Why explicit branches with similar code? Can't all it be handled at once? Something like:
```C#
var left = UnwrapUnary(Visit(mainFromClause.FromExpression));
var right = UnwrapUnary(Visit(containsOperator.Item));
// The constant is on the left side (e.g. db.Users.Where(o => users.Contains(o)))
// The constant is on the right side (e.g. db.Customers.Where(o => o.Orders.Contains(item)))
if (left.NodeType != ExpressionType.Constant && right.NodeType != ExpressionType.Constant)
{
	return false;
}
```

Reply:
Yes it can be handled at once, applied you suggestion.